File: sdks/go/zvault.go
// Package zvault provides the official Go SDK for ZVault Cloud.
//
// Fetch secrets at runtime from ZVault Cloud with in-memory caching,
// auto-refresh, and graceful degradation.
//
// Usage:
//
//	client := zvault.New(os.Getenv("ZVAULT_TOKEN"))
//	secrets, err := client.GetAll(ctx, "production")
//	dbURL := secrets["DATABASE_URL"]
package zvault

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"math/rand/v2"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"
)

const (
	defaultBaseURL    = "https://api.zvault.cloud"
	defaultCacheTTL   = 5 * time.Minute
	defaultTimeout    = 10 * time.Second
	defaultMaxRetries = 3
	retryBaseDelay    = 500 * time.Millisecond
	userAgent         = "zvault-go-sdk/0.1.0"
)

// Config holds configuration for the ZVault client.
type Config struct {
	// Token is the service token (zvt_...) or auth token.
	// Falls back to ZVAULT_TOKEN env var.
	Token string

	// BaseURL is the ZVault Cloud API base URL.
	// Falls back to ZVAULT_URL env var, then https://api.zvault.cloud.
	BaseURL string

	// OrgID is the organization ID.
	// Falls back to ZVAULT_ORG_ID env var.
	OrgID string

	// ProjectID is the project ID.
	// Falls back to ZVAULT_PROJECT_ID env var.
	ProjectID string

	// DefaultEnv is the default environment slug.
	// Falls back to ZVAULT_ENV env var, then "development".
	DefaultEnv string

	// CacheTTL is how long secrets are cached in memory.
	// Default: 5 minutes. Set to 0 to disable caching.
	CacheTTL time.Duration

	// Timeout is the HTTP request timeout. Default: 10 seconds.
	Timeout time.Duration

	// MaxRetries is the number of retry attempts on transient errors.
	// Default: 3.
	MaxRetries int

	// HTTPClient is an optional custom HTTP client. If nil, a default is created.
	HTTPClient *http.Client
}

// SecretEntry represents a single secret from the API.
type SecretEntry struct {
	Key       string `json:"key"`
	Value     string `json:"value"`
	Version   int    `json:"version"`
	Comment   string `json:"comment"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
}

// SecretKey represents a secret key (no value) from list operations.
type SecretKey struct {
	Key       string `json:"key"`
	Version   int    `json:"version"`
	Comment   string `json:"comment"`
	UpdatedAt string `json:"updated_at"`
}

// HealthStatus is the result of a health check.
type HealthStatus struct {
	OK            bool
	LatencyMs     int64
	CachedSecrets int
	LastRefresh   time.Time
}

type secretResponse struct {
	Secret SecretEntry `json:"secret"`
}

type secretKeysResponse struct {
	Keys []SecretKey `json:"keys"`
}

type apiErrorBody struct {
	Error *struct {
		Code    int    `json:"code"`
		Message string `json:"message"`
	} `json:"error"`
}

type cacheEntry struct {
	secrets   map[string]string
	expiresAt time.Time
}

// Client is the ZVault SDK client.
type Client struct {
	token      string
	baseURL    string
	orgID      string
	projectID  string
	defaultEnv string
	cacheTTL   time.Duration
	maxRetries int
	httpClient *http.Client

	mu          sync.RWMutex
	cache       map[string]*cacheEntry // env -> cached secrets
	lastRefresh time.Time
}

// New creates a new ZVault client with the given token.
// Additional configuration can be set via NewWithConfig.
func New(token string) *Client {
	return NewWithConfig(Config{Token: token})
}

// NewWithConfig creates a new ZVault client with full configuration.
func NewWithConfig(cfg Config) *Client {
	token := firstNonEmpty(cfg.Token, os.Getenv("ZVAULT_TOKEN"))
	if token == "" {
		panic("zvault: missing token — set ZVAULT_TOKEN env var or pass Config.Token")
	}

	baseURL := firstNonEmpty(cfg.BaseURL, os.Getenv("ZVAULT_URL"), defaultBaseURL)
	baseURL = strings.TrimRight(baseURL, "/")

	orgID := firstNonEmpty(cfg.OrgID, os.Getenv("ZVAULT_ORG_ID"))
	projectID := firstNonEmpty(cfg.ProjectID, os.Getenv("ZVAULT_PROJECT_ID"))
	defaultEnv := firstNonEmpty(cfg.DefaultEnv, os.Getenv("ZVAULT_ENV"), "development")

	cacheTTL := cfg.CacheTTL
	if cacheTTL == 0 {
		cacheTTL = defaultCacheTTL
	}

	timeout := cfg.Timeout
	if timeout == 0 {
		timeout = defaultTimeout
	}

	maxRetries := cfg.MaxRetries
	if maxRetries == 0 {
		maxRetries = defaultMaxRetries
	}

	httpClient := cfg.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{Timeout: timeout}
	}

	return &Client{
		token:      token,
		baseURL:    baseURL,
		orgID:      orgID,
		projectID:  projectID,
		defaultEnv: defaultEnv,
		cacheTTL:   cacheTTL,
		maxRetries: maxRetries,
		httpClient: httpClient,
		cache:      make(map[string]*cacheEntry),
	}
}

// GetAll fetches all secrets for an environment. Results are cached in-memory.
// On network failure, returns last-known cached values (graceful degradation).
// Pass empty string for env to use the default environment.
func (c *Client) GetAll(ctx context.Context, env string) (map[string]string, error) {
	env = c.resolveEnv(env)
	if err := c.requireProjectConfig(); err != nil {
		return nil, err
	}

	// Fetch key list
	var keysResp secretKeysResponse
	path := fmt.Sprintf("/orgs/%s/projects/%s/envs/%s/secrets", c.orgID, c.projectID, env)
	if err := c.request(ctx, http.MethodGet, path, nil, &keysResp); err != nil {
		// Graceful degradation
		if cached := c.getCached(env); cached != nil {
			return cached, nil
		}
		return nil, err
	}

	// Fetch each secret value
	secrets := make(map[string]string, len(keysResp.Keys))
	for _, k := range keysResp.Keys {
		var resp secretResponse
		secretPath := fmt.Sprintf("%s/%s", path, url.PathEscape(k.Key))
		if err := c.request(ctx, http.MethodGet, secretPath, nil, &resp); err != nil {
			continue // skip individual failures
		}
		secrets[resp.Secret.Key] = resp.Secret.Value
	}

	// Update cache
	c.setCache(env, secrets)
	return secrets, nil
}

// Get fetches a single secret by key. Checks cache first.
// Pass empty string for env to use the default environment.
func (c *Client) Get(ctx context.Context, key, env string) (string, error) {
	env = c.resolveEnv(env)
	if err := c.requireProjectConfig(); err != nil {
		return "", err
	}

	// Check cache
	if val := c.getCachedKey(env, key); val != "" {
		return val, nil
	}

	var resp secretResponse
	path := fmt.Sprintf("/orgs/%s/projects/%s/envs/%s/secrets/%s",
		c.orgID, c.projectID, env, url.PathEscape(key))
	if err := c.request(ctx, http.MethodGet, path, nil, &resp); err != nil {
		return "", err
	}

	// Cache the individual value
	c.setCachedKey(env, key, resp.Secret.Value)
	return resp.Secret.Value, nil
}

// ListKeys lists secret keys (no values) for an environment.
func (c *Client) ListKeys(ctx context.Context, env string) ([]SecretKey, error) {
	env = c.resolveEnv(env)
	if err := c.requireProjectConfig(); err != nil {
		return nil, err
	}

	var resp secretKeysResponse
	path := fmt.Sprintf("/orgs/%s/projects/%s/envs/%s/secrets", c.orgID, c.projectID, env)
	if err := c.request(ctx, http.MethodGet, path, nil, &resp); err != nil {
		return nil, err
	}
	return resp.Keys, nil
}

// Set creates or updates a secret. Requires write permission.
func (c *Client) Set(ctx context.Context, key, value, env, comment string) (*SecretEntry, error) {
	env = c.resolveEnv(env)
	if err := c.requireProjectConfig(); err != nil {
		return nil, err
	}

	body := map[string]string{"value": value, "comment": comment}
	var resp secretResponse
	path := fmt.Sprintf("/orgs/%s/projects/%s/envs/%s/secrets/%s",
		c.orgID, c.projectID, env, url.PathEscape(key))
	if err := c.request(ctx, http.MethodPut, path, body, &resp); err != nil {
		return nil, err
	}

	c.setCachedKey(env, key, value)
	return &resp.Secret, nil
}

// Delete removes a secret. Requires write permission.
func (c *Client) Delete(ctx context.Context, key, env string) error {
	env = c.resolveEnv(env)
	if err := c.requireProjectConfig(); err != nil {
		return err
	}

	path := fmt.Sprintf("/orgs/%s/projects/%s/envs/%s/secrets/%s",
		c.orgID, c.projectID, env, url.PathEscape(key))
	return c.request(ctx, http.MethodDelete, path, nil, nil)
}

// InjectIntoEnv sets all secrets as OS environment variables.
// Existing vars are NOT overwritten unless overwrite is true.
// Returns the number of variables injected.
func (c *Client) InjectIntoEnv(ctx context.Context, env string, overwrite bool) (int, error) {
	secrets, err := c.GetAll(ctx, env)
	if err != nil {
		return 0, err
	}

	count := 0
	for k, v := range secrets {
		if !overwrite && os.Getenv(k) != "" {
			continue
		}
		if err := os.Setenv(k, v); err != nil {
			return count, fmt.Errorf("zvault: failed to set env var %s: %w", k, err)
		}
		count++
	}
	return count, nil
}

// Healthy checks if the ZVault API is reachable and the token is valid.
func (c *Client) Healthy(ctx context.Context) HealthStatus {
	start := time.Now()
	err := c.request(ctx, http.MethodGet, "/me", nil, nil)

	c.mu.RLock()
	cached := 0
	for _, entry := range c.cache {
		if time.Now().Before(entry.expiresAt) {
			cached += len(entry.secrets)
		}
	}
	lastRefresh := c.lastRefresh
	c.mu.RUnlock()

	return HealthStatus{
		OK:            err == nil,
		LatencyMs:     time.Since(start).Milliseconds(),
		CachedSecrets: cached,
		LastRefresh:   lastRefresh,
	}
}

// --- Private methods ---

func (c *Client) resolveEnv(env string) string {
	if env == "" {
		return c.defaultEnv
	}
	return env
}

func (c *Client) requireProjectConfig() error {
	if c.orgID == "" {
		return fmt.Errorf("zvault: missing orgID — set ZVAULT_ORG_ID env var or Config.OrgID")
	}
	if c.projectID == "" {
		return fmt.Errorf("zvault: missing projectID — set ZVAULT_PROJECT_ID env var or Config.ProjectID")
	}
	return nil
}

func (c *Client) getCached(env string) map[string]string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	entry, ok := c.cache[env]
	if !ok || time.Now().After(entry.expiresAt) {
		return nil
	}
	// Return a copy
	result := make(map[string]string, len(entry.secrets))
	for k, v := range entry.secrets {
		result[k] = v
	}
	return result
}

func (c *Client) getCachedKey(env, key string) string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	entry, ok := c.cache[env]
	if !ok || time.Now().After(entry.expiresAt) {
		return ""
	}
	return entry.secrets[key]
}

func (c *Client) setCache(env string, secrets map[string]string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.cache[env] = &cacheEntry{
		secrets:   secrets,
		expiresAt: time.Now().Add(c.cacheTTL),
	}
	c.lastRefresh = time.Now()
}

func (c *Client) setCachedKey(env, key, value string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	entry, ok := c.cache[env]
	if !ok || time.Now().After(entry.expiresAt) {
		entry = &cacheEntry{
			secrets:   make(map[string]string),
			expiresAt: time.Now().Add(c.cacheTTL),
		}
		c.cache[env] = entry
	}
	entry.secrets[key] = value
}

func (c *Client) request(ctx context.Context, method, path string, body any, result any) error {
	fullURL := c.baseURL + "/v1/cloud" + path

	var bodyReader io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return fmt.Errorf("zvault: failed to marshal request body: %w", err)
		}
		bodyReader = strings.NewReader(string(data))
	}

	var lastErr error
	for attempt := 0; attempt <= c.maxRetries; attempt++ {
		req, err := http.NewRequestWithContext(ctx, method, fullURL, bodyReader)
		if err != nil {
			return fmt.Errorf("zvault: failed to create request: %w", err)
		}

		req.Header.Set("Authorization", "Bearer "+c.token)
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("User-Agent", userAgent)

		resp, err := c.httpClient.Do(req)
		if err != nil {
			lastErr = fmt.Errorf("zvault: request failed: %w", err)
			if attempt < c.maxRetries {
				sleepWithJitter(ctx, attempt)
				// Reset body reader for retry
				if body != nil {
					data, _ := json.Marshal(body)
					bodyReader = strings.NewReader(string(data))
				}
				continue
			}
			return lastErr
		}

		defer resp.Body.Close()
		respBody, err := io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("zvault: failed to read response: %w", err)
		}

		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			if result != nil && len(respBody) > 0 {
				if err := json.Unmarshal(respBody, result); err != nil {
					return fmt.Errorf("zvault: failed to parse response: %w", err)
				}
			}
			return nil
		}

		// Parse error
		var apiErr apiErrorBody
		_ = json.Unmarshal(respBody, &apiErr)
		msg := fmt.Sprintf("HTTP %d", resp.StatusCode)
		if apiErr.Error != nil && apiErr.Error.Message != "" {
			msg = apiErr.Error.Message
		}

		// Non-retryable errors
		switch resp.StatusCode {
		case http.StatusUnauthorized, http.StatusForbidden:
			return &APIError{StatusCode: resp.StatusCode, Message: msg}
		case http.StatusNotFound:
			return &APIError{StatusCode: resp.StatusCode, Message: msg}
		}

		// Retryable
		lastErr = &APIError{StatusCode: resp.StatusCode, Message: msg}
		if attempt < c.maxRetries && isRetryable(resp.StatusCode) {
			sleepWithJitter(ctx, attempt)
			if body != nil {
				data, _ := json.Marshal(body)
				bodyReader = strings.NewReader(string(data))
			}
			continue
		}

		return lastErr
	}

	return lastErr
}

// APIError represents an error from the ZVault API.
type APIError struct {
	StatusCode int
	Message    string
}

func (e *APIError) Error() string {
	return fmt.Sprintf("zvault: API error %d: %s", e.StatusCode, e.Message)
}

func isRetryable(status int) bool {
	return status == 429 || status == 500 || status == 502 || status == 503 || status == 504
}

func sleepWithJitter(ctx context.Context, attempt int) {
	delay := retryBaseDelay * time.Duration(math.Pow(2, float64(attempt)))
	jitter := time.Duration(rand.Int64N(int64(float64(delay) * 0.3)))
	select {
	case <-time.After(delay + jitter):
	case <-ctx.Done():
	}
}

func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

File: sdks/elixir/lib/zvault.ex
defmodule ZVault do
  @moduledoc """
  ZVault SDK for Elixir.

  Fetches secrets from ZVault Cloud at runtime. Caches in-memory via ETS,
  auto-refreshes on TTL, and gracefully degrades if the API is unreachable.

  ## Installation

  Add to `mix.exs`:

      def deps do
        [{:zvault, "~> 0.1.0"}]
      end

  ## Usage

      # Start the client (usually in Application supervisor)
      {:ok, _pid} = ZVault.start_link(token: System.get_env("ZVAULT_TOKEN"))

      # Fetch all secrets
      {:ok, secrets} = ZVault.get_all("production")
      db_url = Map.get(secrets, "DATABASE_URL")

      # Fetch single secret
      {:ok, value} = ZVault.get("STRIPE_KEY", "production")

      # List keys
      {:ok, keys} = ZVault.list_keys("production")

      # Health check
      {:ok, health} = ZVault.healthy()
  """

  use GenServer

  @default_base_url "https://api.zvault.cloud"
  @default_cache_ttl 300_000
  @default_timeout 10_000
  @default_max_retries 3
  @retry_base_delay 500

  # --- Public API ---

  def start_link(opts \\ []) do
    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
  end

  @doc "Fetch all secrets for an environment. Returns `{:ok, %{key => value}}`."
  def get_all(env \\ nil) do
    GenServer.call(__MODULE__, {:get_all, env}, 30_000)
  end

  @doc "Fetch a single secret by key. Returns `{:ok, value}` or `{:error, :not_found}`."
  def get(key, env \\ nil) do
    GenServer.call(__MODULE__, {:get, key, env}, 15_000)
  end

  @doc "Set a secret value. Returns `{:ok, secret}`."
  def set(key, value, env \\ nil, comment \\ "") do
    GenServer.call(__MODULE__, {:set, key, value, env, comment}, 15_000)
  end

  @doc "Delete a secret. Returns `:ok`."
  def delete(key, env \\ nil) do
    GenServer.call(__MODULE__, {:delete, key, env}, 15_000)
  end

  @doc "List secret keys (no values). Returns `{:ok, [%{key: key, ...}]}`."
  def list_keys(env \\ nil) do
    GenServer.call(__MODULE__, {:list_keys, env}, 15_000)
  end

  @doc "Check API health. Returns `{:ok, %{ok: bool, latency_ms: int}}`."
  def healthy do
    GenServer.call(__MODULE__, :healthy, 15_000)
  end

  @doc "Inject all secrets into System environment."
  def inject_into_env(env \\ nil) do
    case get_all(env) do
      {:ok, secrets} ->
        count =
          Enum.reduce(secrets, 0, fn {k, v}, acc ->
            System.put_env(k, v)
            acc + 1
          end)

        {:ok, count}

      error ->
        error
    end
  end

  # --- GenServer callbacks ---

  @impl true
  def init(opts) do
    token = opts[:token] || System.get_env("ZVAULT_TOKEN") || ""
    base_url = opts[:base_url] || System.get_env("ZVAULT_URL") || @default_base_url
    org_id = opts[:org_id] || System.get_env("ZVAULT_ORG_ID") || ""
    project_id = opts[:project_id] || System.get_env("ZVAULT_PROJECT_ID") || ""
    default_env = opts[:default_env] || System.get_env("ZVAULT_ENV") || "development"
    cache_ttl = opts[:cache_ttl] || @default_cache_ttl
    timeout = opts[:timeout] || @default_timeout
    max_retries = opts[:max_retries] || @default_max_retries
    debug = opts[:debug] || false

    if token == "" do
      {:stop, "Missing token. Set ZVAULT_TOKEN or pass token: option."}
    else
      cache_table = :ets.new(:zvault_cache, [:set, :private])

      state = %{
        token: token,
        base_url: String.trim_trailing(base_url, "/"),
        org_id: org_id,
        project_id: project_id,
        default_env: default_env,
        cache_ttl: cache_ttl,
        timeout: timeout,
        max_retries: max_retries,
        debug: debug,
        cache: cache_table,
        last_refresh: nil
      }

      {:ok, state}
    end
  end

  @impl true
  def handle_call({:get_all, env}, _from, state) do
    env = env || state.default_env

    case fetch_all_secrets(state, env) do
      {:ok, secrets} ->
        cache_put_all(state.cache, env, secrets, state.cache_ttl)
        state = %{state | last_refresh: DateTime.utc_now()}
        {:reply, {:ok, secrets}, state}

      {:error, _reason} = err ->
        cached = cache_get_all(state.cache, env)

        if map_size(cached) > 0 do
          log(state, "API unreachable, serving #{map_size(cached)} cached secrets")
          {:reply, {:ok, cached}, state}
        else
          {:reply, err, state}
        end
    end
  end

  @impl true
  def handle_call({:get, key, env}, _from, state) do
    env = env || state.default_env

    case cache_get(state.cache, env, key) do
      {:ok, value} ->
        {:reply, {:ok, value}, state}

      :miss ->
        case fetch_secret(state, env, key) do
          {:ok, value} ->
            cache_put(state.cache, env, key, value, state.cache_ttl)
            {:reply, {:ok, value}, state}

          {:error, _} = err ->
            {:reply, err, state}
        end
    end
  end

  @impl true
  def handle_call({:set, key, value, env, comment}, _from, state) do
    env = env || state.default_env
    path = secret_path(state, env, key)
    body = Jason.encode!(%{value: value, comment: comment})

    case request(state, "PUT", path, body) do
      {:ok, %{"secret" => secret}} ->
        cache_put(state.cache, env, key, value, state.cache_ttl)
        {:reply, {:ok, secret}, state}

      {:error, _} = err ->
        {:reply, err, state}
    end
  end

  @impl true
  def handle_call({:delete, key, env}, _from, state) do
    env = env || state.default_env
    path = secret_path(state, env, key)

    case request(state, "DELETE", path) do
      {:ok, _} -> {:reply, :ok, state}
      {:error, _} = err -> {:reply, err, state}
    end
  end

  @impl true
  def handle_call({:list_keys, env}, _from, state) do
    env = env || state.default_env
    path = secrets_path(state, env)

    case request(state, "GET", path) do
      {:ok, %{"keys" => keys}} -> {:reply, {:ok, keys}, state}
      {:error, _} = err -> {:reply, err, state}
    end
  end

  @impl true
  def handle_call(:healthy, _from, state) do
    start = System.monotonic_time(:millisecond)

    result =
      case request(state, "GET", "/me") do
        {:ok, _} ->
          {:ok,
           %{
             ok: true,
             latency_ms: System.monotonic_time(:millisecond) - start,
             last_refresh: state.last_refresh
           }}

        {:error, _} ->
          {:ok,
           %{
             ok: false,
             latency_ms: System.monotonic_time(:millisecond) - start,
             last_refresh: state.last_refresh
           }}
      end

    {:reply, result, state}
  end

  # --- Private helpers ---

  defp secrets_path(state, env) do
    "/orgs/#{state.org_id}/projects/#{state.project_id}/envs/#{env}/secrets"
  end

  defp secret_path(state, env, key) do
    "#{secrets_path(state, env)}/#{URI.encode(key)}"
  end

  defp fetch_all_secrets(state, env) do
    case request(state, "GET", secrets_path(state, env)) do
      {:ok, %{"keys" => keys}} ->
        secrets =
          keys
          |> Enum.map(fn %{"key" => k} -> k end)
          |> Enum.chunk_every(20)
          |> Enum.flat_map(fn batch ->
            batch
            |> Enum.map(fn key ->
              Task.async(fn -> {key, fetch_secret(state, env, key)} end)
            end)
            |> Enum.map(&Task.await(&1, state.timeout + 1_000))
          end)
          |> Enum.filter(fn {_k, result} -> match?({:ok, _}, result) end)
          |> Enum.into(%{}, fn {k, {:ok, v}} -> {k, v} end)

        {:ok, secrets}

      {:error, _} = err ->
        err
    end
  end

  defp fetch_secret(state, env, key) do
    case request(state, "GET", secret_path(state, env, key)) do
      {:ok, %{"secret" => %{"value" => value}}} -> {:ok, value}
      {:error, _} = err -> err
    end
  end

  defp request(state, method, path, body \\ nil) do
    url = "#{state.base_url}/v1/cloud#{path}"

    headers = [
      {"authorization", "Bearer #{state.token}"},
      {"content-type", "application/json"},
      {"user-agent", "zvault-elixir/0.1.0"}
    ]

    do_request_with_retry(method, url, headers, body, state.max_retries, 0, state)
  end

  defp do_request_with_retry(_method, _url, _headers, _body, max_retries, attempt, _state)
       when attempt > max_retries do
    {:error, :max_retries_exceeded}
  end

  defp do_request_with_retry(method, url, headers, body, max_retries, attempt, state) do
    req_opts = [recv_timeout: state.timeout]

    result =
      case method do
        "GET" -> HTTPoison.get(url, headers, req_opts)
        "PUT" -> HTTPoison.put(url, body || "", headers, req_opts)
        "POST" -> HTTPoison.post(url, body || "", headers, req_opts)
        "DELETE" -> HTTPoison.delete(url, headers, req_opts)
      end

    case result do
      {:ok, %HTTPoison.Response{status_code: code, body: resp_body}} when code in 200..299 ->
        if code == 204 do
          {:ok, nil}
        else
          case Jason.decode(resp_body) do
            {:ok, decoded} -> {:ok, decoded}
            {:error, _} -> {:ok, resp_body}
          end
        end

      {:ok, %HTTPoison.Response{status_code: code}} when code in [401, 403] ->
        {:error, {:auth_error, code}}

      {:ok, %HTTPoison.Response{status_code: 404}} ->
        {:error, :not_found}

      {:ok, %HTTPoison.Response{status_code: code}} when code in [429, 500, 502, 503, 504] ->
        if attempt < max_retries do
          delay = @retry_base_delay * :math.pow(2, attempt) |> round()
          jitter = :rand.uniform(round(delay * 0.3))
          log(state, "Retry #{attempt + 1}/#{max_retries} after #{delay + jitter}ms (#{code})")
          Process.sleep(delay + jitter)
          do_request_with_retry(method, url, headers, body, max_retries, attempt + 1, state)
        else
          {:error, {:api_error, code}}
        end

      {:error, %HTTPoison.Error{reason: reason}} ->
        if attempt < max_retries do
          delay = @retry_base_delay * :math.pow(2, attempt) |> round()
          Process.sleep(delay)
          do_request_with_retry(method, url, headers, body, max_retries, attempt + 1, state)
        else
          {:error, {:network_error, reason}}
        end
    end
  end

  defp cache_put(table, env, key, value, ttl) do
    expires_at = System.monotonic_time(:millisecond) + ttl
    :ets.insert(table, {{env, key}, value, expires_at})
  end

  defp cache_put_all(table, env, secrets, ttl) do
    expires_at = System.monotonic_time(:millisecond) + ttl

    Enum.each(secrets, fn {k, v} ->
      :ets.insert(table, {{env, k}, v, expires_at})
    end)
  end

  defp cache_get(table, env, key) do
    case :ets.lookup(table, {env, key}) do
      [{{^env, ^key}, value, expires_at}] ->
        if System.monotonic_time(:millisecond) < expires_at do
          {:ok, value}
        else
          :ets.delete(table, {env, key})
          :miss
        end

      [] ->
        :miss
    end
  end

  defp cache_get_all(table, env) do
    now = System.monotonic_time(:millisecond)

    :ets.foldl(
      fn
        {{^env, key}, value, expires_at}, acc when expires_at > now ->
          Map.put(acc, key, value)

        _, acc ->
          acc
      end,
      %{},
      table
    )
  end

  defp log(%{debug: true}, message) do
    IO.puts("[zvault-sdk] #{message}")
  end

  defp log(_, _), do: :ok
end
